Add a generic cache template library

pranith opened the pull request:
Add a generic cache template to construct internal storage structures. Also add some example use cases by converting the prefetcher tables to use this new library.

pranith committed:
mem-cache: Implement VFT tables using cache library

The frequency table is better built using the generic cache library instead of the
AssociativeSet since the secure bit is not needed for this structure.

Change-Id: Ie3b6442235daec7b350c608ad1380bed58f5ccf4

pranith committed:
mem-cache: prefetch: Implement DCPT tables using cache library

The DCPT table is better built using the generic cache library since we do not
need the secure bit.

Change-Id: I8a4a8d3dab7fbc3bbc816107492978ac7f3f5934

pranith committed:
mem-cache: Derive tagged entry from cache entry

The tagged entry can be derived from the generic cache entry and add the secure
flag that it needs. This reduces code duplication.

Change-Id: I7ff0bddc40604a8a789036a6300eabda40339a0f

giactra commented:
This is a great contribution! It has been in my TODO list for a long time!
I will try to find some time later this week for a review ;)

powerjg commented on src/mem/cache/tags/indexing_policies/IndexingPolicies.py:
```diff
@@ -44,6 +44,9 @@ class BaseIndexingPolicy(SimObject):
     assoc = Param.Int(Parent.assoc, "associativity")
 
+    # number of bits in the tag
+    tag_bits = Param.Unsigned(64, "number of bits in the tag")
```
I think I understand the intention here. However, I still think we need to add more details to the help message.

Maybe "number of bits in the address that are considered part of the tag, index, or offset. If it's less than 64 then the top x = 64 - `tag_bits` are ignored when calculating the tag."

pranith replied:
Updated. Please let me know what you think.

odanrc replied:
Can this new parameter be added in another PR instead? This PR currently does:
- Creates a new non-secure cache (1)
- Updates the prefetchers to use the new cache (1)
- Updates the btbs to use the new cache (2)
- Adds a new parameter to indexing policy (3)

I have added numbers as a suggestion to split this PR up into multiple PRs. The idea is to make it easier to understand which functions were added to the cache because of the prefetcher caches, and which were added because of the BTB, so that we can try to come up with a good API.

pranith replied:
As discussed, it is not easy to split the PR and still be buildable on github. The way to do it is as we do in this PR.

powerjg commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,347 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __CACHE_LIBRARY_HH__
+#define __CACHE_LIBRARY_HH__
```
Usually we use the path and the file name here. E.g., `__BASE_ASSOCIATIVE_CACHE_HH__`

pranith replied:
Done.

powerjg commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,347 @@
+ * Copyright (c) 2024 Pranith Kumar
+ * All rights reserved
+ *
+ * The license below extends only to copyright in the software and shall
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
```
If possible, it would be better to remove this "extra" clause and use the BSD license in the LICENSE file in the root directory.

pranith replied:
Done.

powerjg commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,347 @@
+
+#include "base/cache_entry.hh"
+#include "base/intmath.hh"
+#include "mem/cache/replacement_policies/base.hh"
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
```
After I read through the whole change, I still think this point is important. However, I think you should ignore it for now. Moving around the `ReplacementPolicy` etc. should be done in a separate change and probably needs some discussion. Ignore what I say below, but I'm going to keep it for reference 😄 

If at all possible, I would like to remove cohesion between `src/base` and anything else the codebase. I know that it's not that way right now, but adding more circular dependences is going to make it more difficult to create a more modular build system.

Maybe all of this should be in base?

pranith replied:
Tried doing this. Consensus seems like we might want to do this at a later point.

powerjg commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,347 @@
+
+    void initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
```
There's a lot of code in this file. I suggest splitting out into a .cc and .hh file for maintainability.

pranith replied:
I tried that approach and it was a bit messy due to having to instantiate the template separately for each use case: https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl

I'd much rather prefer this way, but if there is any way to avoid errors  please let me know. Here is the WIP patch doing this (does not compile, has errors): https://github.com/pranith/gem5/commit/143aed0899c4401e3c7bd31d4c9712d0d998fd25

powerjg replied:
Oh, right. I missed that this was a template class. please ignore this comment :)

powerjg reviewed (commented):
Wow, this is surprisingly simple! I really like the change!

We need to ensure that the changes (e.g., store set, btb, etc.) are carefully tested to make sure the functionality doesn't change, but overall, this looks good to me. Really, I only have style comments.

odanrc commented:
A dupm of my understanding of the secure bit: it should be interpreted as part of the address itself, not as a separate entity. If we are using secure spaces, their respective tags **must** make that distinction. Forgetting to assign the secure bits when extracting tags of a secure addr is a bug that has happened in the past. As such, I think that the ideal approach would be to apply design patterns and obj-oriented methodologies to remove the code duplication (e.g., having a polymorphic Addr class that can have both a base Address and SecureAddress counterparts, as well as a matcher function). In this case, for example, getTag would return Addr*, and the code would modified accordingly. Obviously, as mentioned before, this change would require a complete overhaul of the code, and impose performance overheads, so it is not desired. As a consequence, the approach you took in CacheEntry is probably the most reasonable.

src/base/cache_entry.hh extracts the non-secure-bit-related stuff from src/mem/cache/tags/tagged_entry.hh as a new class; the main difference seems to be that most (there are still some, BTW - you should sed there) mentions to the secure bit were removed.  There is one bug with the current implementation, though: the users of TaggedEntry must not be able to access the non-secure functions of CacheEntry. This should be fixed.

src/base/associative_set.hh file is mostly a copy of src/mem/cache/prefetch/associative_set.hh, also removing mentions to the secure bit; however, unlike TaggedEntry the code was just duplicated, and the original file was not modified. This should be fixed to avoid duplicated code.

Finally, since both of these files' contents were copied over from existing files, the original copyright holders **must** be kept.

BTW, TaggedEntry's name no longer makes sense in this context (it should have been SecureEntry or something), but this change is out of the scope of this PR (and it would require going through a deprecation process).

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+
+  protected:
+    /** Name of the cache. */
+    std::string cache_name;
```
Remove and make this class inherit from base/named.hh

pranith replied:
Done.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+    size_t numEntries;
+
+    /** Associativity of the cache. */
+    size_t _associativity;
```
Keep the standard of the file: remove the underscore

pranith replied:
Done.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+    BaseIndexingPolicy *indexingPolicy;
+
+    /** Number of sets in the cache. */
+    size_t numSets;
```
This belongs to the indexing policy - it should be removed from here.

pranith replied:
Removed.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+        initParams(numEntries, _associativity);
+    }
+
+    ~AssociativeCache() {
```
Nit: you can default this destructor, and it should do the correct steps

pranith replied:
That is clearly much cleaner. Done.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+     * Return the name of the cache
+     * @return name of the cache
+     */
+    std::string name() const { return cache_name; }
```
Remove (with Named)

pranith replied:
Added derivation from Named() and removed this.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+     * Clear the entries in the cache.
+     */
+    void clear() {
+        for (auto set = 0; set < numSets; set++) {
```
for (auto &entry : entries) {
    invalidate(&entry);
}

pranith replied:
ah, much simpler...

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+     * @param addr Addr to check for
+     * @return if a valid entry exists
+     */
+    virtual bool isEntryValid(const Addr addr)
```
This should be removed from the API, as it is mostly synthetic sugar. Makes the learning curve easier, as there are less functions to memorize (and maintain).

pranith replied:
ok. removed it.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,338 @@
+     * @param addr Addr to get the tag for
+     * @return Tag for the address
+     */
+    virtual size_t getTag(const Addr addr) const
```
The type of a tag is not size_t. It is either an Addr (preferred), or a new Tag type. Also, is this function really needed? I need to check later

pranith replied:
Updated it to use Addr.

odanrc commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -89,21 +77,20 @@ class TaggedEntry : public ReplaceableEntry
 
     /** Invalidate the block. Its contents are no longer valid. */
-    virtual void invalidate()
+    virtual void invalidate() override
```
Remove the virtual keyword

pranith replied:
removed.

odanrc reviewed (commented):
I did a very quick review of 2 files and I will dump the feedback for now. I will get back to the review later.

pranith commented:
Thank you for taking the time for reviewing the patches.

> 
> src/base/cache_entry.hh extracts the non-secure-bit-related stuff from src/mem/cache/tags/tagged_entry.hh as a new class; the main difference seems to be that most (there are still some, BTW - you should sed there) mentions to the secure bit were removed. There is one bug with the current implementation, though: the users of TaggedEntry must not be able to access the non-secure functions of CacheEntry. This should be fixed.

Updated by adding definitions that 'panic' if used without 'is_secure' arg.

> 
> src/base/associative_set.hh file is mostly a copy of src/mem/cache/prefetch/associative_set.hh, also removing mentions to the secure bit; however, unlike TaggedEntry the code was just duplicated, and the original file was not modified. This should be fixed to avoid duplicated code.

This part is the next item I will tackle. I tried doing this, but derivig from a template and create another template needs some kind of template specialization and I was running into compilation errors. It does not seem as straight forward.

> 
> Finally, since both of these files' contents were copied over from existing files, the original copyright holders **must** be kept.

Done.

giactra commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,339 @@
+        replPolicy->reset(entry->replacementData);
+    }
+
+    /**
+     * Find the set of entries that could be replaced given
+     * that we want to add a new entry with the provided key
+     * @param addr key to select the set of entries
+     * @result vector of candidates matching with the provided key
+     */
+    std::vector<Entry *>
+    getPossibleEntries(const Addr addr) const
+    {
+        std::vector<ReplaceableEntry *> selected_entries =
+            indexingPolicy->getPossibleEntries(addr);
+        std::vector<Entry *> entries(selected_entries.size(), nullptr);
+
+        unsigned int idx = 0;
+        for (auto &entry : selected_entries) {
+            entries[idx++] = static_cast<Entry *>(entry);
+        }
+
+        return entries;
+    }
```
Hi, I would love this method to not be "confined" to the use of replaceable entries only if possible

The Arm TLB supports partial translation entries, meaning a cached translation could store an intermediate step in the translation walk. This leads to a TLB that stores more than one valid entry.. Needless to say the lookup logic prioritizes complete translations over partial ones.. For this reason we would make use of the getPossibleEntries over the findEntry, which returns a single entry only and returns the first entry that matches the TAG....

pranith replied:
Hi @giactra, does marking the method virtual help? That way you can modify it to your specific needs in a derived class. Here 'Entry' has to be derived from 'ReplaceableEntry' for any of the replacement policies to work. I am not sure how we can remove the restriction, please let me know if you have any suggestions. Thanks!

pranith committed:
mem-cache: Implement AssociativeSet from AssociativeCache

AssociativeSet can reuse most of the generic cache library code with the
addition of a secure bit. This reduces duplicated code.

Change-Id: I008ef79b0dd5f95418a3fb79396aeb0a6c601784

pranith commented on src/mem/cache/prefetch/stride.cc:
```diff
@@ -106,14 +106,15 @@ Stride::allocateNewContext(int context)
-        PCTable(pcTableInfo.assoc, pcTableInfo.numEntries,
-        pcTableInfo.indexingPolicy, pcTableInfo.replacementPolicy,
-        StrideEntry(initConfidence))));
+	new PCTable("PCTable", pcTableInfo.numEntries, pcTableInfo.assoc,
```
I am not sure why, but the destructor for PCTable was being called without this 'new'. Any ideas why? How was it working before?

odanrc replied:
Because pcTables is a map of <int, PCTable>. It is not a pointer, so it shouldn´t be newed.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+  protected:
+
+    /** The number of entries in the cache. */
+    size_t numEntries;
```
I would remove this variable. It is a duplicate of entries.size once the cache is initialized.

pranith replied:
Done.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+
+#include <vector>
+
+#include "base/cache_entry.hh"
```
Include:
<type_traits> for std::is_base_of
base/logging.hh for fatal_if

pranith replied:
Done.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+    /**
+     * Empty constructor - need to call init() later with all args
+     */
+    AssociativeCache(const char *name) : Named(std::string(name)) {}
```
Why do we want this constructor? I imagine there is an use case in this PR, so can you point it out for me? Ideally we don´t want to leave room for a user forgetting to call init(); the API should ideally be created in a way that minimizes the likelihood of bugs.

BTW, this constructor can probably be defaulted.

pranith replied:
This is because of structures like MemDep that use init() to initialize their members. I think the flexibility is useful.

https://github.com/gem5/gem5/pull/745/commits/ad48c30bc8a886bba8cd486739f6042f7bf2d2ac#diff-9f51ee10311c0bfdad6797c70f108373a6ddffdeb6fbd9140cd1a690078a3410R100

pranith replied:
@odanrc Any other input before resolving this comment?

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+    /**
+     * Clear the entries in the cache.
+     */
+    void clear() {
```
The following is not a strong opinion, so you can opt to ignore the suggested changes:

clear, in std, typically (always?) means removing all elements from the container - i.e., zeroing its size. Ideally we should mimic std´s nomenclature, as users are used to it, so the learning curve is smoother. I would rename it as invalidateBlocks or something instead. However, we already use ¨insert¨ in a different fashion from std, so I guess it doesn´t matter that much in the end.

In mem/cache/tags/base you will find a forEachBlk function. It applies a function to all blocks. I think it could be better to have an equivalent iterator function here that the user could apply any function to all entries, including invalidation.

pranith replied:
I like the suggestion. I am open to renaming both insert/clear() if it makes it more consistent.

pranith replied:
@odanrc Do you suggest I change the naming to insert/clear()?

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+     */
+    virtual Entry* findEntry(const Addr addr) const
+    {
+        auto tag   = getTag(addr);
```
Remove extra spaces

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+              BaseIndexingPolicy *_indexing_policy,
+              Entry const &init_val = Entry())
+    {
+        numEntries           = num_entries;
```
I think aligning the = is not part of the coding style, but I am not sure

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+        auto entries = indexingPolicy->getPossibleEntries(addr);
+
+        for (auto it = entries.begin(); it != entries.end(); it++) {
+            Entry *entry = static_cast<Entry*>(*it);
```
I *think* you can replace lines 194-197 by something more concise:
        for (auto &entry : entries) {
            if (entry.matchTag(tag)) {
                return &entry;

pranith replied:
The entries here are vector of pointers. Renamed it to 'candidates' to make it clearer. We still need the static_cast<> since the container is of 'ReplaceableEntry*' and matchTag() does not exist in that class. Updated accordingly.

odanrc commented on src/base/associative_cache.hh:
```diff
@@ -0,0 +1,308 @@
+     * @result vector of candidates matching with the provided key
+     */
+    std::vector<Entry *>
+    getPossibleEntries(const Addr addr) const
```
Why is this new function being added? I don´t think it is being used anywhere in this PR. If this function is eventually needed to make something else use the cache (e.g., what Giacomo requests), it should be added in the respective PR that makes use of it. This approach of not adding until is needed is the standard to avoid having stale/forgotten code.

pranith replied:
I moved it to the commit which makes use of this function (changing AssociativeSet to use AssociativeCache).

odanrc commented on src/base/cache_entry.hh:
```diff
@@ -0,0 +1,135 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __CACHE_ENTRY_HH__
```
`__BASE_CACHE_ENTRY_HH__`

pranith replied:
Done.

odanrc commented on src/base/cache_entry.hh:
```diff
@@ -0,0 +1,135 @@
+     * @sa invalidate()
+     * @sa insert()
+     */
+    bool _valid;
```
`bool valid{false};` (notice the removal of the underscore too)

odanrc commented on src/base/cache_entry.hh:
```diff
@@ -0,0 +1,135 @@
+    bool _valid;
+
+    /** The entry's tag. */
+    Addr _tag;
```
`Addr tag{MaxAddr};` (ditto)

pranith replied:
Done.

odanrc commented on src/base/cache_entry.hh:
```diff
@@ -0,0 +1,135 @@
+class CacheEntry : public ReplaceableEntry
+{
+  public:
+    CacheEntry() : ReplaceableEntry(), _valid(false), _tag(MaxAddr) {}
```
With the changes I suggest below this can be defaulted.

odanrc commented on src/base/cache_entry.hh:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    /** Invalidate the block. Its contents are no longer valid. */
+    virtual void invalidate()
```
New line after void

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -29,6 +29,7 @@
 #ifndef __CACHE_PREFETCH_ASSOCIATIVE_SET_HH__
 #define __CACHE_PREFETCH_ASSOCIATIVE_SET_HH__
 
+#include "base/associative_cache.hh"
```
Include type_traits for is_base_of

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -83,28 +81,7 @@ class AssociativeSet
-     * @result vector of candidates matching with the provided key
-     */
-    std::vector<Entry *> getPossibleEntries(const Addr addr) const;
+    using AssociativeCache<Entry>::findEntry;
```
We don´t want to be able to use that one here, since it lacks the secure bit

pranith replied:
Moved it to 'private' so that users cannot invoke this method.

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -113,55 +90,11 @@ class AssociativeSet
      * @param entry pointer to the container entry to be inserted
      */
     void insertEntry(Addr addr, bool is_secure, Entry* entry);
+    using AssociativeCache<Entry>::insertEntry;
```
We don´t want to be able to use that one here, since it lacks the secure bit

pranith replied:
Moved it to private.

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -113,55 +90,11 @@ class AssociativeSet
-    {
-        return entries.end();
-    }
+  private:
```
Why are these 3 lines needed?

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -42,38 +43,35 @@ namespace gem5
      */
-    AssociativeSet(int assoc, int num_entries, BaseIndexingPolicy *idx_policy,
-        replacement_policy::Base *rpl_policy, Entry const &init_val = Entry());
+    AssociativeSet(const char *name, const size_t num_entries,
```
`= default` or using AssociativeCache´s constructor

pranith replied:
Tried making it default, but I am getting an error that says 'only special member functions may be defaulted'.

odanrc commented on src/mem/cache/prefetch/associative_set_impl.hh:
```diff
@@ -35,93 +35,28 @@
-        Entry* entry = static_cast<Entry *>(location);
-        if ((entry->getTag() == tag) && entry->isValid() &&
-            entry->isSecure() == is_secure) {
+    for (auto it = entries.begin(); it != entries.end(); it++) {
```
Ditto (what I said for lines 194-197 of the other file)

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/indirect_memory.cc:
```diff
@@ -44,11 +44,15 @@ IndirectMemory::IndirectMemory(const IndirectMemoryPrefetcherParams &p)
+                  p.pt_table_indexing_policy,
                   PrefetchTableEntry(p.num_indirect_counter_bits)),
-    ipd(p.ipd_table_assoc, p.ipd_table_entries, p.ipd_table_indexing_policy,
+    ipd("IPD", p.ipd_table_entries, p.ipd_table_assoc,
```
Why do we want the caches to be named? Also, as it is the names are not unique. Shouldn´t we use the name of the SimObject instantiating the cache to form a unique name?

pranith replied:
The name is so that if we want to DPRINT() etc., we can have a proper log of which cache is generating this output. I agree that we should append this name to the name of the parent SimObject.

odanrc replied:
Bump

pranith replied:
I am adding the name of the SimObject to the objects name so that they are not unique.

odanrc commented on src/mem/cache/prefetch/stride.hh:
```diff
@@ -137,7 +137,7 @@ class Stride : public Queued
     };
     typedef AssociativeSet<StrideEntry> PCTable;
-    std::unordered_map<int, PCTable> pcTables;
+    std::unordered_map<int, PCTable*> pcTables;
```
This explains why new is now needed. Why was this change to pointer needed?

pranith replied:
Without new, the make_pair() call is calling the desctructor of the PCTable() object. I think the map containers insert()/make_pair() invokes the copy/move constructor that is calling destructors.

pranith replied:
There seems to be a benign bug in the existing code base. I added a destructor to the AssociativeSet as follows:

~AssociativeSet() = default;

And the simulation crashes immediately after creating and inserting the PCTable(). Converting it to a pointer seems to fix it.

pranith replied:
@odanrc Any input for this comment?

odanrc replied:
This change should be reverted

pranith replied:
Can you please explain why? Not using a pointer was causing the object to be constructed and then copied.

odanrc replied:
Bump

pranith replied:
Can you please explain why you want this change reverted? I think the change is necessary.

odanrc replied:
Maybe it is because the copy and assignment operators were removed from AssociativeCache (just a guess). Nonetheless, I think this is a bug that is introduced by this PR, and that it needs to be fixed in order for this PR to be merged.

pranith replied:
I think using a pointer is actually better than copying the entire array elements for this case. Can you please check this updated code where we are saving a pointer to the table instead of copying it?

odanrc replied:
There was only one place initializing the table:
`pcTables.insert(std::make_pair(context, PCTable(`

After the table is constructed, it is not deleted nor overwritten, and there are possibly one or two static copies (if, and only if, make_pair or insert trigger one). However, with pointers we end up with a frequent dereference overhead. I would say that frequent overhead is likely worse than a single one upon initialization.

odanrc replied:
In any case, when a small side change generates a discussion that makes the review longer, if possible IMHO it is better to drop it from the main PR and make a proposal PR just for the side change in order to get the main PR going. I believe this is my last remaining request, but I will double check over the weekend.

pranith replied:
> There was only one place initializing the table: `pcTables.insert(std::make_pair(context, PCTable(`
> 
> After the table is constructed, it is not deleted nor overwritten, and there are possibly one or two static copies (if, and only if, make_pair or insert trigger one). However, with pointers we end up with a frequent dereference overhead. I would say that frequent overhead is likely worse than a single one upon initialization.

I don't quite agree with the overhead statement here. Saving a pointer and referencing the newly constructed table in the map is as efficient as referencing an object within the container. And we actually do want a pointer as you can see in the findTable() function, it returns a pointer eventually.

But that discussion is not that relevant for the current patch, so I reverted it to what it was previously and am storing the PCTable object in the map using piecewise constructor so that we don't have to copy the object immediately after creating it.

odanrc commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,42 +101,28 @@ class TaggedEntry : public ReplaceableEntry
-     * @param tag The tag value.
-     */
-    virtual void setTag(Addr tag) { _tag = tag; }
+    using CacheEntry::matchTag;
```
Isn´t needing lines 105-108 a consequence of adding line 104? If you remove lines 104-108 won´t the base function be inaccessible?

pranith replied:
No, without adding the 'using...' I get an error message that the newly added 'matchTag' hides an overloaded virtual function in the base class.

pranith replied:
@odanrc Any new input for this comment before resolving?

odanrc replied:
Bump

pranith replied:
I replied to your comment about why this is necessary. Let me know if it did not answer your question.

odanrc replied:
Would it be possible to mark these as private then?

pranith replied:
Done. Although I am not sure how this works, since we have the same matchTag with different arguments. How do we mark one without secure tag private and make the one with is_secure tag to be public.

odanrc commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,42 +101,28 @@ class TaggedEntry : public ReplaceableEntry
+        return false;
+    }
 
+    using CacheEntry::insert;
```
Ditto (lines 113-117 is a consequence of the addition of line 112)

odanrc replied:
Bump

pranith replied:
See above.

odanrc reviewed (commented):
I haven´t had time to review the changes in cpu, but here is another review dump.

pranith commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -113,55 +90,11 @@ class AssociativeSet
-    {
-        return entries.end();
-    }
+  private:
```
Without adding 'using rp/ip.' I get an error that the IP and RP pointers do not exist. I think this is something new with the recent compilers.

Similarly, if I don't add 'using findEntry', I get an error that the newly added 'findEntry' with the additional 'is_secure' argument hides overloaded virtual function.

pranith replied:
Done.

andysan commented:
Thanks for doing this, this change is long overdue! I won't have time to review this properly, but I have a high-level design suggestions. It seems like we currently assume that tags have the type `Addr`, I think this might be undesirable. There are many cases where we need a lot more data than a simple address in a tag. An obvious example is a TLB where the tag may contain things like an ASID, VMID, S/NS, and parts of the address (see for example `src/dev/arm/smmu_v3_caches.hh`). Similarly, you could imagine BTB entries being tagged with ASID and VMID.

Similarly, I think it would be desirable to not make as many assumptions about the type used lookups. An obvious example of this is the way we currently handle the NS/S bit which ends up being passed separately as a special case. There may well be more of these bits in the future. 

It would be a huge improvement if you could make the tag and lookup keys template parameters, but I don't know if this would be feasible in this PR though. It might have a large knock-on effect on the indexing policies.

I think what we need to truly generalise this is a way to specify the following:

- A lookup type. This could be a `struct` or `Addr`
- A hash function that works on the lookup type. Modern caches don't always use the text-book shift and mask approach. This may need to be a parameter.
- A comparison operator. In practice, we should probably store the entire lookup type as a tag instead of throwing away bits to save space. This would effectively require an `operator==` for the lookup type which would just work for many types.

As a concrete example, this would let you define a cache using the following pseudo-code:
```
struct CacheKey {
    uint16_t vmid;
    uint16_t asid;
    bool ns;
    Addr pa;
};

uint64_t
my_hash(const CacheKey &key)
{
    return (key.pa >> bits) & mask_bits;
}

AssociativeCache<CacheKey, my_hash> my_cache;
```

You'd then perform lookups using instances of `CacheKey`.

Again, you don't necessarily need to implement this in this PR. The current refactor already provides significant benefit. It would however be nice we could generalise it properly once and for all to avoid API churn.

powerjg commented:
@pranith can you create (or link) an issue which describes the work that you're doing. There are many PRs (some open, some closed), and I am having a hard time keeping track. Also, the rate at which you're pushing new changes is much higher than the rate at which we can review changes. Given this mismatch of bandwidths 😉 it's going to be difficult for us to get these changes merged. I really like the direction, and I think it's a good contribution to gem5, but we need to make sure that we take it step by step and ensure that it's as easy as possible for the community to review the changes before they can be merged.

pranith commented:
Thank you for the detailed comments @andysan.

> 
> Again, you don't necessarily need to implement this in this PR. The current refactor already provides significant benefit. It would however be nice we could generalise it properly once and for all to avoid API churn.

I will create a discussion with the points you raised so that we can plan and work on it. I think its a better idea to make minimal changes to the current PR to ease review and merging @powerjg pointed out.

pranith commented:
> @pranith can you create (or link) an issue which describes the work that you're doing. There are many PRs (some open, some closed), and I am having a hard time keeping track. Also, the rate at which you're pushing new changes is much higher than the rate at which we can review changes. Given this mismatch of bandwidths 😉 it's going to be difficult for us to get these changes merged. I really like the direction, and I think it's a good contribution to gem5, but we need to make sure that we take it step by step and ensure that it's as easy as possible for the community to review the changes before they can be merged.

Hi @powerjg, this PR is the master for adding the cache template. All the changes I am making are in response to the helpful review comments by @odanrc to make the code more suitable for merging. The new PRs I opened were suggested by @odanrc to ease review. There is no easier way to break up this PR since most of the later commits depend on the previous ones. Please let me know if there is anything I can do to ease review. I can understand that this is a large change and will take time to review properly. I will keep my changes to a minimum and address the feedback and not make any new ones. Thanks!

odanrc commented:
As mentioned before, I'm very much in favor of what @andysan proposes; however, i agree that if you decide to apply that, it should probably be done in a PR later down the chain,  since the indexing policies will likely need to be modified, and using templates in SimObjects is usually not trivial.

@powerjg I think it is easier to review the broken down PR, which starts with https://github.com/gem5/gem5/pull/770

@pranith please ping me once you have updated that PR with the changes I requested here.

odanrc reviewed (commented):
I disagree with moving caches to the base directory. I understand that indexing and replacement policies should be used by all caches, and that caches can be used within other SimObjects (i.e., not the memory´s cache hierarchy), but for me a cache is still a object with a memory functionality, so it makes sense to be in the mem directory.

odanrc commented:
Commit https://github.com/gem5/gem5/pull/745/commits/b7a1064bb971fe1efacd1224a1dd49fac6342a6f (mem-cache: Add a tag mask to indexing policies) should not be in this PR. If it is needed to rework some of the objects to use the new cache, then such change needs its own PR.

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -62,20 +63,35 @@ struct ltseqnum
+        SSITEntry() : CacheEntry(), _ssid(MaxAddr) {}
+
+        void setSSID(SSID id) { _ssid = id; }
+        SSID getSSID(void) { return _ssid; }
```
Should be a const function

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -149,16 +158,14 @@ class StoreSet
 
     /** Store Set ID Table size, in entries. */
     int SSITSize;
+    int SSITAssoc;
```
Remove this variable

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -149,16 +158,14 @@ class StoreSet
     int SSITSize;
+    int SSITAssoc;
+
+    BaseReplacementPolicy *SSITReplPolicy;
```
These 2 variables belong in the cache, not the user of the cache. Remove them from here

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -62,20 +63,35 @@ struct ltseqnum
   public:
-    typedef unsigned SSID;
+    typedef Addr SSID;
+    typedef replacement_policy::Base BaseReplacementPolicy;
```
Remove this typedef, as it costs 61 characters to add it, but it currently (after my requested changes) saves 4*3=12 characters. One would have to use this defined type quite a lot of times to be worth it.

pranith replied:
I am not too against removing this. But its more of a style preference rather than saving characters typed. We are mixing 'replacement_policy::Base' and 'BaseIndexingPolicy'. Having this typedef keeps the code more consistent when reading.

pranith replied:
This is done.

odanrc reviewed (changes requested):
Review of commit https://github.com/gem5/gem5/pull/745/commits/2fb2ddab225998621d8698b409288cebd8ac85eb

odanrc commented on src/cpu/pred/btb_indexing_policy.hh:
```diff
@@ -0,0 +1,116 @@
+     * @param tid The thread id
+     * @return The possible entries.
+     */
+    std::vector<ReplaceableEntry*>
```
Move implementation to .cc

odanrc commented on src/cpu/pred/btb_indexing_policy.hh:
```diff
@@ -0,0 +1,116 @@
+        return sets[extractSet(instPC, tid)];
+    }
+
+    void setNumThreads(unsigned num_threads)
```
Move implementation to .cc

odanrc commented on src/cpu/pred/btb_indexing_policy.hh:
```diff
@@ -0,0 +1,116 @@
+
+} // namespace gem5
+
+#endif //__MEM_CACHE_INDEXING_POLICIES_SET_ASSOCIATIVE_HH__
```
Fix header guard

odanrc commented on src/cpu/pred/btb_indexing_policy.hh:
```diff
@@ -0,0 +1,116 @@
+ * Declaration of a BTB indexing policy.
+ */
+
+#ifndef __BTB_INDEXING_POLICY_HH__
```
Fix header guard (`__CPU_PRED_BTB_INDEXING_POLICY_HH__`)

odanrc commented on src/cpu/pred/btb_indexing_policy.cc:
```diff
@@ -0,0 +1,65 @@
+BTBIndexingPolicy::BTBIndexingPolicy(const Params &p)
+    : SetAssociative(p)
+{
+    log2NumThreads = log2i(p.numThreads);
```
Call setNumThreads instead

odanrc commented on src/cpu/pred/btb_indexing_policy.hh:
```diff
@@ -0,0 +1,116 @@
+{
+
+class ReplaceableEntry;
+
```
Put this class in namespace branch_prediction

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
-                           StaticInstPtr inst = nullptr) override;
-    const StaticInstPtr getInst(ThreadID tid, Addr instPC) override;
+    /** The entry's target. */
+    std::unique_ptr<PCStateBase> target;
```
Include memory

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+    /** Match the tag of the BTB entry */
+    bool matchTag(const Addr _tag, ThreadID _tid)
+    {
+        return (isValid() && (tag == _tag) && (tid == _tid));
```
Call CacheEntry::matchTag(_tag)

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+    }
+
+    /** Set the tag of the BTB entry */
+    void setTag(Addr _tag) override { tag = _tag; }
```
Remove

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
   public:
-    SimpleBTB(const SimpleBTBParams &params);
+    /** The entry's tag. */
+    Addr tag;
```
Remove (already in CacheEntry)

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other)
+    {
+        tag   = other.tag;
```
Call CacheEntry´s copy constr

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+    /** Default constructor */
+    BTBEntry() : CacheEntry(), tag(0), tid(0), valid(false), inst(nullptr)  {}
+
+    using CacheEntry::matchTag;
```
Why do we allow both to be used? Don´t we always need to know the thread id? Also, is the thread id part of the tag (i.e., can we have multiple threads using the same tag)? If so, then the API of the cache should use a Tag type, not only a regular addr.

pranith replied:
Without adding this 'using...' line, the compiler complains that the newly added 'matchTag' hides an overloaded virtual function in the base class.

pranith replied:
@odanrc Does the reply satisfy your question? If so, please resolve this.

odanrc replied:
Shouldn´t this using be marked private?

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+{
   private:
-    struct BTBEntry
+    typedef replacement_policy::Base BaseReplacementPolicy;
```
Remove (re-implementing AssociativeCache)

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
-    struct BTBEntry
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+    BTBIndexingPolicy *indexingPolicy;
```
Remove (re-implementing AssociativeCache)

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+    BTBIndexingPolicy *indexingPolicy;
+
+  public:
+    BTBCache(const char *_my_name, const size_t num_entries,
```
Move all implementation to .cc

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +54,137 @@ namespace gem5
+        indexingPolicy = static_cast<BTBIndexingPolicy*>(_indexingPolicy);
+    }
+
+    BTBEntry *findEntry(const Addr instPC, ThreadID tid)
```
We are duplicating a lot of the code in the associative cache just because (it seems) the thread id is part of the address. All of these overrides would be avoided if we used a Tag type instead in the associative cache instead.

pranith replied:
Agree, but that is a much bigger change where we also have to touch indexing policies to use the Tag type. I prefer we do it later after these changes.

pranith commented:
> I disagree with moving caches to the base directory. I understand that indexing and replacement policies should be used by all caches, and that caches can be used within other SimObjects (i.e., not the memory´s cache hierarchy), but for me a cache is still a object with a memory functionality, so it makes sense to be in the mem directory.

We are not moving caches to the base directory. Just the IPs and RPs that are generic and can be used with the associative cache in subsystems other than the memory subsystem. It looks off to include 'mem/cache/tags/indexing_policies' in a frontend or LS file.

This change was initially suggested by @powerjg  and then I moved them to 'base/cache' after your suggestion in PR #760. I am OK either way but lets please decide one way or the other because this take quite a bit of time to move.

pranith commented:
> Commit [b7a1064](https://github.com/gem5/gem5/commit/b7a1064bb971fe1efacd1224a1dd49fac6342a6f) (mem-cache: Add a tag mask to indexing policies) should not be in this PR. If it is needed to rework some of the objects to use the new cache, then such change needs its own PR.

I agree that tag mask should be in its own PR (PR #769), because it can be merged on its own without any build issues. However, any change that uses the new cache needs to be in the same PR to build. As we discussed before, I don't see how to make one PR depend on a previous PR to avoid build issues. What is the advantage that you see with independent PRs that do not build?

pranith commented:
Hi @odanrc, is there any other review feedback for this change? Please let me know how you want to proceed to get this changed merged. Thank you!

powerjg commented:
> > I disagree with moving caches to the base directory. I understand that indexing and replacement policies should be used by all caches, and that caches can be used within other SimObjects (i.e., not the memory´s cache hierarchy), but for me a cache is still a object with a memory functionality, so it makes sense to be in the mem directory.
> 
> We are not moving caches to the base directory. Just the IPs and RPs that are generic and can be used with the associative cache in subsystems other than the memory subsystem. It looks off to include 'mem/cache/tags/indexing_policies' in a frontend or LS file.
> 
> This change was initially suggested by @powerjg and then I moved them to 'base/cache' after your suggestion in PR #760. I am OK either way but lets please decide one way or the other because this take quite a bit of time to move.

If Daniel or others would prefer the indexing policies, etc. not moved to base, that's fine with me. After we add the generic cache template we can think about how we want to move thing around.

> I agree that tag mask should be in its own PR (PR https://github.com/gem5/gem5/pull/769), because it can be merged on its own without any build issues. However, any change that uses the new cache needs to be in the same PR to build. As we discussed before, I don't see how to make one PR depend on a previous PR to avoid build issues. What is the advantage that you see with independent PRs that do not build?

This is just a difference in how github and gerrit works. There are a few cases where gerrit is more natural, but we moved to github PRs because that's where most of the world is now. If you want to have changesets with "dependencies" then, that means there should be a PR with multiple commits (i.e., like this).

pranith commented:
> If Daniel or others would prefer the indexing policies, etc. not moved to base, that's fine with me. After we add the generic cache template we can think about how we want to move thing around.
> 

Updated the PR removing the move of IPs and RPs.

> This is just a difference in how github and gerrit works. There are a few cases where gerrit is more natural, but we moved to github PRs because that's where most of the world is now. If you want to have changesets with "dependencies" then, that means there should be a PR with multiple commits (i.e., like this).

Great. May be we should add it to documentation on contributing? Thanks!

odanrc commented:
> Hi @odanrc, is there any other review feedback for this change? Please let me know how you want to proceed to get this changed merged. Thank you!

Hey, Pranith! I haven´t reviewed this PR again because there are still multiple comments/questions I made pending action/answer. Could you tackle those?

By the way, because the of the rename from mem/cache/* to base/* and back, the includes got out of lexicographical order. I don´t recall how to fix that quickly, but there should be an automated way in the codebase to do that to the commits in this PR (probably using the pre-commit file).

pranith committed:
misc: Add a generic cache library

Add a generic cache library modeled after AssociativeSet that can be used for
constructing internal caching structures.

Change-Id: I1767309ed01f52672b32810636a09142ff23242f

pranith commented:
> Hey, Pranith! I haven´t reviewed this PR again because there are still multiple comments/questions I made pending action/answer. Could you tackle those?

I updated the comments with my answers and resolved the ones I fixed.

> 
> By the way, because the of the rename from mem/cache/* to base/* and back, the includes got out of lexicographical order. I don´t recall how to fix that quickly, but there should be an automated way in the codebase to do that to the commits in this PR (probably using the pre-commit file).

Yeesh, that was a mess. Cleaned it up now. Please take a look. Thanks!

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
+    StaticInstPtr inst;
+
+    /** Update the entry in the BTB */
+    void update(ThreadID _tid,
```
Move the implementation to .cc

pranith replied:
Done.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
-        /** The entry's target. */
-        std::unique_ptr<PCStateBase> target;
+    /** Match the tag of the BTB entry */
+    bool matchTag(const Addr _tag, ThreadID _tid)
```
Move the implementation to .cc

pranith replied:
Done.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
-        /** The entry's thread id. */
-        ThreadID tid;
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other) : CacheEntry(other)
```
Move the implementation to .cc

pranith replied:
Done.

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
+    }
+
+    /** Default constructor */
+    BTBEntry() : CacheEntry(), tid(0), inst(nullptr)  {}
```
Move the implementation to .cc

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+namespace gem5
+{
+
+/** Cache Library.
```
This comment is not really helping the reader, since one can extract the exact same information just from reading the class/function name. Could you either improve (preferable), or drop it?

pranith replied:
I couldn't think of anything better, so I removed it.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+
+  private:
+
+    void initParams(size_t _num_entries, size_t _assoc)
```
Since we are doing the implementation here, we should follow the implementation´s guidelines. The return type should be in its own line. i.e.:
```
void
initParams(...)
{
```

More info on the coding style can be found here: https://www.gem5.org/documentation/general_docs/development/coding_style/

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * Clear the entries in the cache.
+     */
+    void clear() {
```
Ditto (return type in its own line, and open bracket on the line below clear)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    void init(const size_t num_entries,
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * @param addr Addr to get the tag for
+     * @return Tag for the address
+     */
+    virtual Addr getTag(const Addr addr) const
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * @param addr key to the entry
+     * @return The entry if it exists
+     */
+    virtual Entry* accessEntry(const Addr addr)
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * Update the replacement information for an entry
+     * @param Entry to access and upate
+     */
+    virtual void accessEntry(Entry *entry)
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * @return returns a pointer to the wanted entry or nullptr if it does not
+     *  exist.
+     */
+    virtual Entry* findEntry(const Addr addr) const
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     */
+    virtual Entry* findEntry(const Addr addr) const
+    {
+        auto tag   = getTag(addr);
```
Nit: Remove extra spaces between tag and =

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * @param addr key to select the possible victim
+     * @result entry to be victimized
+     */
+    virtual Entry* findVictim(const Addr addr)
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     *
+     * @param entry Entry to be invalidated.
+     */
+    virtual void invalidate(Entry *entry)
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * @param addr key of the container
+     * @param entry pointer to the container entry to be inserted
+     */
+    virtual void insertEntry(Addr addr, Entry *entry)
```
Ditto (return type line)

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * Returns an iterator to the first entry of the dictionary
+     * @result iterator to the first element
+     */
+    iterator begin()
```
Ditto (return type line)

pranith replied:
Done.

odanrc replied:
Bump

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * (placeholder element, should not be accessed)
+     * @result iterator to the end element
+     */
+    iterator end()
```
Ditto (return type line)

pranith replied:
Done.

odanrc replied:
Bump

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * Returns an iterator to the first entry of the dictionary
+     * @result iterator to the first element
+     */
+    const_iterator begin() const
```
Ditto (return type line)

pranith replied:
Done.

odanrc replied:
Bump

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+     * (placeholder element, should not be accessed)
+     * @result iterator to the end element
+     */
+    const_iterator end() const
```
Ditto (return type line)

pranith replied:
Done.

odanrc replied:
Bump

pranith replied:
Done.

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,315 @@
+    {
+        std::vector<ReplaceableEntry *> selected_entries =
+            indexingPolicy->getPossibleEntries(addr);
+        std::vector<Entry *> entries(selected_entries.size(), nullptr);
```
Nit: This is an std-y version of this code (include algorithm):
```
std::vector<ReplaceableEntry *> selected_entries =
    indexingPolicy->getPossibleEntries(addr);
std::vector<Entry *> entries;
std::transform(selected_entries.begin(), selected_entries.end(), 
    std::back_inserter(entries), [](auto &entry){
        return static_cast<Entry *>(entry);
    });
return entries;
```

pranith replied:
That is neat, thanks! updated.

odanrc commented on src/cpu/o3/store_set.cc:
```diff
@@ -82,7 +76,9 @@ StoreSet::~StoreSet()
 void
-StoreSet::init(uint64_t clear_period, int _SSIT_size, int _LFST_size)
+StoreSet::init(uint64_t clear_period, int _SSIT_size, int _SSIT_assoc,
+               replacement_policy::Base *_replPolicy,
```
_replPolicy and _indexingPolicy are not following the pseudo snake case aprroach taken here. They should be _repl_policy and _indexing_policy, respectively

odanrc commented on src/cpu/o3/BaseO3CPU.py:
```diff
@@ -154,6 +156,13 @@ def support_take_over(cls):
+        LRURP(), "SSIT replacement policy"
+    )
+    SSITIndexingPolicy = Param.BaseIndexingPolicy(
+        SetAssociative(), "SSIT indexing policy"
```
The original code had some assumptions about the indexing policy. For example, the offsetBits are 2, so the entry_size is 4. This change changes the original behavior, which is not desired.

As a major side note, I am not familiar with the code in CPU. In order to receive a proper review, I highly recommend that you remove all changes made to the CPU directory from this PR and leave to do those in another PR; otherwise, a knowledgeable maintainer of this dir should review this section thoroughly (@powerjg ?)

odanrc commented on src/cpu/o3/store_set.cc:
```diff
@@ -39,8 +39,13 @@ namespace gem5
-StoreSet::StoreSet(uint64_t clear_period, int _SSIT_size, int _LFST_size)
-    : clearPeriod(clear_period), SSITSize(_SSIT_size), LFSTSize(_LFST_size)
+StoreSet::StoreSet(uint64_t clear_period, int _SSIT_size, int _SSIT_assoc,
+                   replacement_policy::Base *_replPolicy,
```
Ditto fix naming of the policies to snake case

odanrc commented on src/cpu/o3/store_set.cc:
```diff
@@ -108,79 +99,71 @@ StoreSet::init(uint64_t clear_period, int _SSIT_size, int _LFST_size)
 
     if (!valid_load_SSID && !valid_store_SSID) {
         // Calculate a new SSID here.
         SSID new_set = calcSSID(load_PC);
```
Rename new_set to ssid

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -62,20 +63,34 @@ struct ltseqnum
     /** Creates store set predictor with given table sizes. */
-    StoreSet(uint64_t clear_period, int SSIT_size, int LFST_size);
+    StoreSet(uint64_t clear_period, int SSIT_size, int SSIT_assoc,
+             replacement_policy::Base *replPolicy,
```
Ditto camel case to snake case

odanrc commented on src/cpu/o3/store_set.hh:
```diff
@@ -62,20 +63,34 @@ struct ltseqnum
     /** Initializes the store set predictor with the given table sizes. */
-    void init(uint64_t clear_period, int SSIT_size, int LFST_size);
+    void init(uint64_t clear_period, int SSIT_size, int SSIT_assoc,
+              replacement_policy::Base *_replPolicy,
```
Ditto camel case to snake case

odanrc commented on src/cpu/o3/BaseO3CPU.py:
```diff
@@ -154,6 +156,13 @@ def support_take_over(cls):
     )
     LFSTSize = Param.Unsigned(1024, "Last fetched store table size")
     SSITSize = Param.Unsigned(1024, "Store set ID table size")
```
Are the SSITSize and SSITAssoc used elsewhere in the code, or are they just used to instantiate the indexing policy? If the latter, then I think they should be part of the default instantiation of the indexing policy (line 164). e.g.: SetAssociative(size=1024, assoc=1, entry_size=4)

odanrc commented on src/cpu/pred/BranchPredictor.py:
```diff
@@ -93,6 +103,13 @@ class SimpleBTB(BranchTargetBuffer):
+        LRURP(), "BTB replacement policy"
+    )
+    btbIndexingPolicy = Param.BaseIndexingPolicy(
+        BTBIndexingPolicy(tag_bits=Self.tagBits), "BTB indexing policy"
```
Ditto move associativity to indexing policy

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
+    BTBEntry* findVictim(const Addr addr, ThreadID tid);
 
+  private:
+    using AssociativeCache<BTBEntry>::accessEntry;
```
There are 6 functions in AssociativeCache´s public API that use Addr, yet only 3 of them are hidden away here. This BTB Cache uses the ThreadID as part of the address, if I understand correctly, so the users of this class should never be able to use those functions. The other 3 should be hidden away too.

odanrc replied:
Document why these become private

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
-        StaticInstPtr inst = nullptr;
-    };
+  public:
+    BTBCache(const char *_my_name, const size_t num_entries,
```
Move the implementation of the constructor to the .cc file

odanrc commented on src/cpu/pred/simple_btb.hh:
```diff
@@ -52,41 +56,89 @@ namespace gem5
+class BTBCache : public AssociativeCache<BTBEntry>
+{
+  private:
+    BTBIndexingPolicy *indexingPolicy;
```
Document why, even though we are inheriting from AssociativeCache, which already contains and indexing policy, we are adding another one here. This will allow people to see this as an issue in the future, and possibly commit to fix it.

odanrc commented on src/cpu/pred/BranchPredictor.py:
```diff
@@ -93,6 +103,13 @@ class SimpleBTB(BranchTargetBuffer):
     instShiftAmt = Param.Unsigned(
```
This is entry_size in the indexing policy, so this parameter should be removed and instantiated only there, or it should be used like you do for tagBits (i.e., entry_size=Self.instShiftAmt)

odanrc commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -42,38 +45,32 @@ namespace gem5
      */
-    AssociativeSet(int assoc, int num_entries, BaseIndexingPolicy *idx_policy,
-        replacement_policy::Base *rpl_policy, Entry const &init_val = Entry());
+    AssociativeSet(const char *name, const size_t num_entries,
```
Move the implementation of the constructor back to the _impl.hh file.

odanrc replied:
Bump

odanrc reviewed (changes requested):
I have noticed that there are still multiple comments pending reply. I also saw that some comments were marked as resolved without any respective changes or replies. I would like to ask you to please, when a reviewer asks a question or suggests something to do, explain the reasoning of why no change will be made, as something that is obvious to the author may not be obvious to the reviewer, specially if they wrote a comment/suggestion about it.

pranith commented on src/mem/cache/prefetch/associative_set.hh:
```diff
@@ -42,38 +45,32 @@ namespace gem5
      */
-    AssociativeSet(int assoc, int num_entries, BaseIndexingPolicy *idx_policy,
-        replacement_policy::Base *rpl_policy, Entry const &init_val = Entry());
+    AssociativeSet(const char *name, const size_t num_entries,
```
Done.

pranith replied:
This is already done.

pranith commented:
Hi @odanrc, I updated the branch with a new commit addressing your review comments. I removed the store set and BTB changes so that we can focus on the cache/prefetcher in this PR. I will update those other changes with your review comments and send a new PR once this change is merged. If you are ok with these changes, I will squash the review commit appropriately. Thanks!

odanrc commented:
Hey, Pranith. There are still many pending comments. Please make sure that all comments have been acted upon.

giactra committed:
stdlib: Use newly defined tree for PrivateL1PrivateL2 hierarchy

Change-Id: I803c6118c4df62484018f9e4d995026adb1bbc2c
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
stdlib: Add tree structure to the AbstractCacheHierarchy

One of things we miss in gem5 is the capability to neatly compose the
cache hierarchy of CPUs and clusters of CPUs.  The BaseCPU
addPrivateSplitL1Caches and addTwoLevelCacheHierarchy APIs have
historically been used to bind cache levels together.

These APIs have been superseeded by the introduction of the Cache
hierarchy abstraction in the standard library. The standard library
makes it cleaner for a user to quickly instantiate a hierarchy of caches
with few lines of code.  While this removes a lot of complexity for a
user, the Hierarchy objects still have little information about their
internal topology.

To address this problem, this patch adds a tree data structure to the
AbstractCacheHierarchy class, where every node of the tree represent
a cache in the hierarchy. In this way we will expose APIs for traversing
and querying the tree.

For example a 2 CPUs system with private L1, private L2 and shared L3
will contain the following tree:

         [root]
           |
          [L3]
           /\
          /  \
        [L2] [L2]
         |    |
        [L1] [L1]

Change-Id: I78fe6ad094f0938ff9bed191fb10b9e841418692
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

ivanaamit commented:
Hi @pranith, could you please address the pending comments? Thank you.

pranith commented:
@odanrc The only review comments that I did not address are related to BTB/BranchPredictor/O3CPU. I have removed all the changes related to these in the latest patch so that we can first focus on the cache part. I will send the other changes in a different PR.

I went through all the comments and made sure I resolved all the cache relevant review comments. Please let me know if I missed something. I added a new commit that addresses all the review comments.

Thanks!

odanrc reviewed (changes requested):
@pranith I have bumped the missing comments. In order to see the comments, I click on the `Files Changed` tab (https://github.com/gem5/gem5/pull/745/files) and scroll down.

pranith commented:
> @pranith I have bumped the missing comments. In order to see the comments, I click on the `Files Changed` tab (https://github.com/gem5/gem5/pull/745/files) and scroll down.

Thanks @odanrc. I added comments to your bumps and addressed some of them in the new commit. The rest I think need more explanation and I pointed out which ones. Please let me know. Thanks!

odanrc commented:
@pranith , I cannot see any comments pointed out. Could you make sure that your questions have been sent?

pranith reviewed (commented):
Oh, I did not submit my comments all these days!!

pranith commented:
> @pranith , I cannot see any comments pointed out. Could you make sure that your questions have been sent?

Sorry! I added comments to most of your reviews but did not know that you have to then 'Submit review' for those comments to show up. No wonder I did not get a response. Please check now.

odanrc commented on src/mem/cache/prefetch/pif.cc:
```diff
@@ -46,8 +46,9 @@ PIF::PIF(const PIFPrefetcherParams &p)
       historyBuffer(p.history_buffer_size),
-      index(p.index_assoc, p.index_entries, p.index_indexing_policy,
-            p.index_replacement_policy),
+      index("PIFIndex", p.index_entries, p.index_assoc,
```
Non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/signature_path.cc:
```diff
@@ -48,12 +48,16 @@ SignaturePath::SignaturePath(const SignaturePathPrefetcherParams &p)
-                     p.signature_table_replacement_policy),
-      patternTable(p.pattern_table_assoc, p.pattern_table_entries,
-                   p.pattern_table_indexing_policy,
+      signatureTable("SignatureTable",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/signature_path.cc:
```diff
@@ -48,12 +48,16 @@ SignaturePath::SignaturePath(const SignaturePathPrefetcherParams &p)
+		     p.signature_table_assoc,
+                     p.signature_table_replacement_policy,
+                     p.signature_table_indexing_policy),
+      patternTable("PatternTable",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/irregular_stream_buffer.cc:
```diff
@@ -44,19 +44,23 @@ IrregularStreamBuffer::IrregularStreamBuffer(
-                 p.training_unit_indexing_policy,
-                 p.training_unit_replacement_policy),
-    psAddressMappingCache(p.address_map_cache_assoc,
+    trainingUnit("TrainingUnit",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/irregular_stream_buffer.cc:
```diff
@@ -44,19 +44,23 @@ IrregularStreamBuffer::IrregularStreamBuffer(
+		 p.training_unit_assoc,
+                 p.training_unit_replacement_policy,
+                 p.training_unit_indexing_policy),
+    psAddressMappingCache("PSAddressMappingCache",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/irregular_stream_buffer.cc:
```diff
@@ -44,19 +44,23 @@ IrregularStreamBuffer::IrregularStreamBuffer(
                           AddressMappingEntry(prefetchCandidatesPerEntry,
                                               p.num_counter_bits)),
-    spAddressMappingCache(p.address_map_cache_assoc,
+    spAddressMappingCache("SPAddressMappingCache",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/signature_path_v2.cc:
```diff
@@ -42,10 +42,11 @@ namespace prefetch
 SignaturePathV2::SignaturePathV2(const SignaturePathPrefetcherV2Params &p)
     : SignaturePath(p),
-      globalHistoryRegister(p.global_history_register_entries,
+      globalHistoryRegister("GlobalHistoryRegister",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -42,16 +42,18 @@ STeMS::STeMS(const STeMSPrefetcherParams &p)
     spatialRegionSizeBits(floorLog2(p.spatial_region_size)),
     reconstructionEntries(p.reconstruction_entries),
-    activeGenerationTable(p.active_generation_table_assoc,
+    activeGenerationTable("ActiveGenerationTable",
```
Ditto non-unique name

pranith replied:
Done.

odanrc commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -42,16 +42,18 @@ STeMS::STeMS(const STeMSPrefetcherParams &p)
                           ActiveGenerationTableEntry(
                               spatialRegionSize / blkSize)),
-    patternSequenceTable(p.pattern_sequence_table_assoc,
+    patternSequenceTable("PatternSequenceTable",
```
Ditto non-unique name

pranith replied:
Done.

odanrc reviewed (commented):
This explains why we have been in this feedback limbo for so long. Hopefully we get things going now :)

pranith reviewed (commented):
Updated names of the caches to be unique.

giactra committed:
stdlib: Add DTB generation capabilites to AbstractCacheHierarchy

Now that we are able to provide a view of the cache hierarchy from
the python world, we can start generating DTB entries for caches
and more specifically to properly fill the next-level-cache and
cache-level properties

Change-Id: Iba9ea08fe605f77a353c9e64d62b04b80478b4e2
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

odanrc commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,328 @@
+        auto entry = findEntry(addr);
+
+        if (entry) {
+            replPolicy->touch(entry->replacementData);
```
Nit: call accessEntry(entry) instead

pranith replied:
Updated.

odanrc commented on src/mem/cache/prefetch/delta_correlating_prediction_tables.cc:
```diff
@@ -42,15 +42,15 @@ namespace prefetch
    deltaBits(p.delta_bits), deltaMaskBits(p.delta_mask_bits),
-   table(p.table_assoc, p.table_entries, p.table_indexing_policy,
-         p.table_replacement_policy, DCPTEntry(p.deltas_per_entry))
+   table("DCPT", p.table_entries, p.table_assoc, p.table_replacement_policy,
```
Non-unique name

pranith replied:
Fixed.

odanrc commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -90,15 +92,12 @@ STeMS::checkForActiveGenerationsEnd(const CacheAccessor &cache)
-                ActiveGenerationTableEntry *pst_entry =
-                    patternSequenceTable.findEntry(pst_addr,
-                                                   false /*unused*/);
+                ActiveGenerationTableEntry *pst_entry = patternSequenceTable.findEntry(pst_addr);
```
This line is currently going beyond the character limit

pranith replied:
Fixed.

odanrc commented on src/mem/cache/prefetch/spatio_temporal_memory_streaming.cc:
```diff
@@ -222,8 +221,7 @@ STeMS::reconstructSequence(
         it++) {
-        ActiveGenerationTableEntry *pst_entry =
-            patternSequenceTable.findEntry(it->pstAddress, false /* unused */);
+        ActiveGenerationTableEntry *pst_entry = patternSequenceTable.findEntry(it->pstAddress);
```
Ditto beyond limit

pranith replied:
Fixed.

odanrc commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,44 +102,34 @@ class TaggedEntry : public ReplaceableEntry
-     * @param tag The tag value.
-     */
-    virtual void setTag(Addr tag) { _tag = tag; }
+    bool
```
Is this reachable given the existence of line 131?

pranith replied:
probably not.. but I will try out removing it in another PR.

odanrc commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,44 +102,34 @@ class TaggedEntry : public ReplaceableEntry
+        return false;
+    }
 
+    void
```
Ditto, line 132

odanrc reviewed (approved):
LGTM, with a few details that you can leave for a future PR if desired.

pranith commented:
Do I squash and/or merge the review commits into the appropriate commits?

odanrc commented:
@powerjg , my approval continued despite more changes being made to the PR. Is this intended, or a GitHub misconfig?

(yes, I still approve the PR, but I don´t think the approval should have been automatically kept)

odanrc commented:
@pranith, I don´t know what is the procedure now that we use GitHub. @powerjg @giactra , could you answer Pranith´s question?

powerjg commented:
@pranith 
> Do I squash and/or merge the review commits into the appropriate commits?

If you have more than 1 commit that you think should show up in the commit history, then, yes, please rebase/squash/etc. to make the history for this PR look good. If you have more than 1 commit in the PR, we will do a merge commit. On the other hand, if this PR should be a *single* commit, we can squash and merge. You don't have to do anything in this case.

@odanrc 
> my approval continued despite more changes being made to the PR. Is this intended, or a GitHub misconfig?

I don't really know. I *think* this is just how github works. I believe that Pranith (or a maintainer) would need to "dismiss" your review to force you to re-review. However, if you think I'm wrong, I can dig more into the documentation/options.

pranith commented:
> If you have more than 1 commit that you think should show up in the commit history, then, yes, please rebase/squash/etc. to make the history for this PR look good. If you have more than 1 commit in the PR, we will do a merge commit. On the other hand, if this PR should be a _single_ commit, we can squash and merge. You don't have to do anything in this case.

Yes, I think it is cleaner to have more than one commit. I will merge in the review commits to the appropriate ones and update the PR.

pranith commented:
@powerjg this PR is ready to be merged.

giactra committed:
Update src/python/gem5/components/cachehierarchies/abstract_cache_hierarchy.py

Co-authored-by: Bobby R. Bruce <bbruce@ucdavis.edu>

ivanaamit commented:
@giactra, could you please review this PR? If you don't have time to do a detailed review, we would appreciate it if you could at least approve the general idea and changes. Thank you.

giactra commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,328 @@
+     * @param num_entries total number of entries of the container, the number
+     *   of sets can be calculated dividing this balue by the 'assoc' value
+     * @param associativity number of elements in each associative set
+     * @param entry_size Size of each entry
```
entry_size is not part of the constructor

pranith replied:
Updated.

giactra commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,44 +102,34 @@ class TaggedEntry : public ReplaceableEntry
-     */
-    virtual void setTag(Addr tag) { _tag = tag; }
+    bool
+    matchTag(const Addr tag) const override {
+        panic("Need is_secure arg");
+        return false;
+    }
```
Minor coding style, this should be

bool
matchTag(const Addr tag) const override
{
}

pranith replied:
Updated.

giactra commented on src/mem/cache/tags/tagged_entry.hh:
```diff
@@ -114,44 +102,34 @@ class TaggedEntry : public ReplaceableEntry
+    }
 
+    void
+    insert(const Addr tag) override {
+        panic("Need is_secure arg");
+        return;
+    }
```
Ditto

pranith replied:
Updated.

giactra commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,328 @@
+template <typename Entry>
+class AssociativeCache : public Named
+{
+    static_assert(std::is_base_of_v<CacheEntry, Entry>,
```
I am ok for keeping this for now until the API stabilizes, but as we are using a template parameter (which is good) I am not sure I see a reason for  enforcing the usage of a CacheEntry subclass. We should be able to create a Cache with any interface compliant Entry

pranith replied:
Updated.

giactra commented on src/base/cache/associative_cache.hh:
```diff
@@ -0,0 +1,328 @@
+    void
+    initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
+                 "AssociativeCache<> must be a power of 2");
+        fatal_if(!isPowerOf2(_assoc), "The associativity of an "
+                 "AssociativeCache<> must be a power of 2");
```
This is not always the case. For example the L2 TLB in N1 core is a 5-way set associative 1280 entry [1]. I believe the condition here should be:

    fatal_if(_num_entries % assoc != 0)

[1]: https://en.wikichip.org/wiki/arm_holdings/microarchitectures/neoverse_n1

pranith replied:
Updated.

odanrc replied:
FYI, the indexing policy is implemented with the restriction that the number of sets is a power of 2. Removing this fatal from here is desired, since the AssocCache does not care whether it is a pow of 2 or not; however, to support non-pow2 architectures the indexing policies will need to be slightly reworked (as in adding some assertions and masks here and there, and fixing the tagShift and setShift, possibly fixing the skewed assoc cache). Obviously, this is not a request to do such rework, just an FYI :)

giactra replied:
Thanks for reporting this @odanrc. I will create an Issue detailing this so that it doesn't get lost

giactra reviewed (changes requested):
Hi pranith, I have some comments but then I believe this is good to go.

pranith commented:
> Hi pranith, I have some comments but then I believe this is good to go.

Hi @giactra, I added a commit fixing your review comments. Once you resolve the comments, I will merge them back into appropriate commits. Thanks!

giactra commented:
> > Hi pranith, I have some comments but then I believe this is good to go.
> 
> Hi @giactra, I added a commit fixing your review comments. Once you resolve the comments, I will merge them back into appropriate commits. Thanks!

Done

pranith commented:
> > > Hi pranith, I have some comments but then I believe this is good to go.
> > 
> > 
> > Hi @giactra, I added a commit fixing your review comments. Once you resolve the comments, I will merge them back into appropriate commits. Thanks!
> 
> Done

@giactra This PR is now ready for merging.

giactra reviewed (approved):
LGTM, will merge when tests will finish
